### PR TITLE
Dnsmasq nanny

### PR DESCRIPTION
**Pull request by bowei**
This adds dnsmasq-nanny, a helper process that wraps dnsmasq and configures it in response to the kube-dns configmap. This implements two configuration options:
* stubDomains: a map of additional stub domains that are resolved by outside nameservers
* upstreamNameservers: a list of upstream nameservers to use in place of the one from /etc/resolv.conf

### CODE REVIEW

File: pkg/dnsmasq/nanny_test.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+	for _, tc := range testCases {
+		args := tc.args
+		gomega.Expect(ExtractDnsmasqArgs(&args)).To(
+			gomega.Equal(tc.dnsmasqArgs))
+		gomega.Expect(args).To(gomega.Equal(tc.otherArgs))
+	}
+}
+
+func TestNannyConfig(t *testing.T) {
```
Comment (MrHohn) [inline pkg/dnsmasq/nanny_test.go]: May also use testCases here so it has the same style as the other unit tests.

File: test/fixtures/mock-dnsmasq.sh [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+runForever() {
+	echo "runForever"
+	while true; do
+		sleep 1
+	done
+}
+
+sleepThenError() {
+	echo "exitWithError"
```
Comment (MrHohn) [inline test/fixtures/mock-dnsmasq.sh]: Should be "sleepThenError".

File: cmd/dnsmasq-nanny/main.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+		flag.PrintDefaults()
+	}
+
+	flag.StringVar(&opts.dnsmasqExec, "dnsmasqExec", opts.dnsmasqExec,
+		"location of the dnsmasq executable")
+	flag.StringVar(&opts.configDir, "configDir", opts.configDir,
+		"location of the configuration")
+	flag.BoolVar(&opts.restartDnsmasq, "restartDnsmasq",
+		opts.restartDnsmasq,
+		"location of the dnsmasq executable")
```
Comment (MrHohn) [inline cmd/dnsmasq-nanny/main.go]: Wrong description?

File: cmd/dnsmasq-nanny/main.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+	configChan := sync.Periodic()
+
+	for {
+		select {
+		case status := <-daemon.ExitChannel:
+			glog.Fatalf("dnsmasq exited: %v", status)
+			break
+		case currentConfig = <-configChan:
+			if opts.restartDnsmasq {
```
Comment (MrHohn) [inline cmd/dnsmasq-nanny/main.go]: Wouldn't the nanny keep restarting Dnsmasq everytime it reads the config?

File: cmd/dnsmasq-nanny/main.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+	daemon := &dnsmasq.Nanny{Exec: opts.dnsmasqExec}
+	daemon.Configure(opts.dnsmasqArgs, currentConfig)
+	if err := daemon.Start(); err != nil {
+		glog.Fatalf("Could not start dnsmasq with initial configuration: %v", err)
+	}
+
+	configChan := sync.Periodic()
+
+	for {
```
Comment (MrHohn) [inline cmd/dnsmasq-nanny/main.go]: May be better to put the majority logics of main() into pkg/dnsmasq and make it some function calls here?

File: pkg/dns/config/sync.go [hunk 6]
```diff
@@ -130,18 +145,34 @@ func (sync *kubeSync) processUpdate(result syncResult, buildUnchangedConfig bool
...
+	return nil
+}
+
+func updateStubDomains(key string, value string, config *Config) error {
+	config.StubDomains = make(map[string][]string)
+	if err := json.Unmarshal([]byte(value), &config.StubDomains); err != nil {
+		glog.Errorf("Invalid JSON: %q", value)
+		return err
+	}
+
```
Comment (MrHohn) [inline pkg/dns/config/sync.go]: Should log updateStubDomains() and updateUpstreamNameservers() just like updateFederations().

File: pkg/dns/config/sync.go [hunk 7]
```diff
@@ -130,18 +145,34 @@ func (sync *kubeSync) processUpdate(result syncResult, buildUnchangedConfig bool
...
+	glog.V(2).Infof("Updated federations to %v", config.Federations)
 
-	return
+	return nil
+}
+
+func updateStubDomains(key string, value string, config *Config) error {
+	config.StubDomains = make(map[string][]string)
+	if err := json.Unmarshal([]byte(value), &config.StubDomains); err != nil {
+		glog.Errorf("Invalid JSON: %q", value)
```
Comment (MrHohn) [inline pkg/dns/config/sync.go]: Something like `Invalid JSON for stub domains`?

File: pkg/dns/config/sync.go [hunk 8]
```diff
@@ -130,18 +145,34 @@ func (sync *kubeSync) processUpdate(result syncResult, buildUnchangedConfig bool
...
+		return err
+	}
+
+	return nil
+}
+
+func updateUpstreamNameservers(key string, value string, config *Config) error {
+	if err := json.Unmarshal([]byte(value), &config.UpstreamNameservers); err != nil {
+		glog.Errorf("Invalid JSON: %q", value)
```
Comment (MrHohn) [inline pkg/dns/config/sync.go]: ditto

File: pkg/e2e/dnsmasq/nanny.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+)
+
+type Harness struct {
+	TmpDir      string
+	NannyExec   string
+	MockDnsmasq string
+}
+
+func (h *Harness) Setup() {
+	if err := os.Mkdir(h.TmpDir+"/config", 0755); err != nil {
```
Comment (cmluciano) [inline pkg/e2e/dnsmasq/nanny.go]: Would [TempDir](https://golang.org/pkg/io/ioutil/#TempDir) make this easier?

Comment (bowei) [inline pkg/e2e/dnsmasq/nanny.go]: h.TmpDir is already a temporary directory created by the outer runner...

File: cmd/dnsmasq-nanny/main.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
```
Comment (bowei) [inline cmd/dnsmasq-nanny/main.go]: probably overkill. the daemon is not meant to be run by hand in any case

File: test/e2e/e2e_test.go [hunk 11]
```diff
@@ -41,6 +47,41 @@ func failureHandler(message string, callerSkip ...int) {
 	ginkgo.Fail(message, callerSkip...)
 }
 
+// GinkgoLogger suppresses logging unless "-v" is given to the gingko framework.
+type GinkgoLogger struct{}
```
Comment (bowei) [inline test/e2e/e2e_test.go]: discussed.

File: test/e2e/e2e_test.go [hunk 12]
```diff
@@ -41,6 +47,41 @@ func failureHandler(message string, callerSkip ...int) {
...
+}
+
+func (*GinkgoLogger) LogWithPrefix(prefix string, str string) {
+	if config.DefaultReporterConfig.Verbose {
+		e2e.LogWithPrefix(log.Printf, logPreamble()+prefix, str)
+	}
+}
+
+func logPreamble() (preamble string) {
+	preamble = "\x1b[1mLOG: \x1b[0m"
```
Comment (bowei) [inline test/e2e/e2e_test.go]: fixed

File: pkg/dnsmasq/nanny_test.go (hunk 1 above)
Comment (bowei) [inline pkg/dnsmasq/nanny_test.go]: ok

File: test/fixtures/mock-dnsmasq.sh (hunk 2 above)
Comment (bowei) [inline test/fixtures/mock-dnsmasq.sh]: done

File: cmd/dnsmasq-nanny/main.go (hunk 3 above)
Comment (bowei) [inline cmd/dnsmasq-nanny/main.go]: fixed

File: cmd/dnsmasq-nanny/main.go (hunk 4 above)
Comment (bowei) [inline cmd/dnsmasq-nanny/main.go]: it has to receive a new configuration first...

File: pkg/dns/config/sync.go (hunk 6 above)
Comment (bowei) [inline pkg/dns/config/sync.go]: done

File: pkg/dns/config/sync.go (hunk 7 above)
Comment (bowei) [inline pkg/dns/config/sync.go]: done

**Comment by bowei** [comment]
updated

**Pull request merged**